vernac/declare.ml

```diff
@@ -910,6 +910,10 @@ let declare_definition_core ~info ~cinfo ~opaque ~obls ~body ?using sigma =
 let declare_definition ~info ~cinfo ~opaque ~body ?using sigma =
   declare_definition_core ~obls:[] ~info ~cinfo ~opaque ~body ?using sigma |> fst
 
+let declare_definition_full ~info ~cinfo ~opaque ~body ?using sigma =
+  let c, uctx = declare_definition_core ~obls:[] ~info ~cinfo ~opaque ~body ?using sigma in
+  c, UState.context_set uctx
```

Comment:
IMO the `if poly` should be here rather than in the caller https://github.com/LPCIC/coq-elpi/blob/414b24e3c8a2febc530e379e7f41ab07f18634cd/src/coq_elpi_builtins.ml#L957

Reply:
but I really would rather have the universes_entry I think

Reply:
done, but I did not get your second comment.

Reply:
I mean something like
~~~diff
modified   vernac/declare.ml
@@ -905,14 +905,14 @@ let declare_definition_core ~info ~cinfo ~opaque ~obls ~body ?using sigma =
   let { Info.scope; clearbody; kind; hook; typing_flags; user_warns; ntns; _ } = info in
   let gref = declare_entry_core ~name ~scope ~clearbody ~kind ~impargs ~typing_flags ~user_warns ~obls ?hook ~uctx entry in
   List.iter (Metasyntax.add_notation_interpretation ~local:(info.scope=Locality.Discharge) (Global.env ())) ntns;
-  gref, uctx
+  (gref, fst entry.proof_entry_universes), uctx
  
-let declare_definition ~info ~cinfo ~opaque ~body ?using sigma =
+let declare_definition_full ~info ~cinfo ~opaque ~body ?using sigma =
   declare_definition_core ~obls:[] ~info ~cinfo ~opaque ~body ?using sigma |> fst
 
-let declare_definition_full ~info ~cinfo ~opaque ~body ?using sigma =
-  let c, uctx = declare_definition_core ~obls:[] ~info ~cinfo ~opaque ~body ?using sigma in
-  c, if info.poly then Univ.ContextSet.empty else UState.context_set uctx
+let declare_definition ~info ~cinfo ~opaque ~body ?using sigma =
+  declare_definition_full ~info ~cinfo ~opaque ~body ?using sigma |> fst
 
 let prepare_obligations ~name ?types ~body env sigma =
   let env = Global.env () in
@@ -1347,7 +1347,7 @@ let declare_definition ~pm prg =
   let name, info, opaque, using = prg.prg_cinfo.CInfo.name, prg.prg_info, prg.prg_opaque, prg.prg_using in
   let obls = List.map (fun (id, (_, c)) -> (id, c)) varsubst in
   (* XXX: This is doing normalization twice *)
-  let kn, uctx = declare_definition_core ~cinfo ~info ~obls ~body ~opaque ?using sigma in
+  let (kn, _), uctx = declare_definition_core ~cinfo ~info ~obls ~body ~opaque ?using sigma in
   (* XXX: We call the obligation hook here, by consistency with the
      previous imperative behaviour, however I'm not sure this is right *)
   let pm = State.call_prg_hook prg
modified   vernac/declare.mli
@@ -446,7 +446,7 @@ val declare_definition_full
   -> body:EConstr.t
   -> ?using:Vernacexpr.section_subset_expr
   -> Evd.evar_map
-  -> GlobRef.t * Univ.ContextSet.t
+  -> GlobRef.t * UState.universes_entry
 
 (** Declaration messages, for internal use *)
~~~
this feels more reliable than UState.context_set, also in the poly case callers may want to get the associated instance from the entry.